File: Data Structures & Algorithms/three-integer-sum/submission-10.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        
        res = []
        nums.sort()

        for i, n in enumerate(nums):
            lp = i + 1
            rp = len(nums) - 1

            if n > 0:
                break
            if i > 0 and n == nums[i - 1]:
                continue
            
            while lp < rp:
                if nums[lp] + nums[rp] + n < 0:
                    lp += 1
                elif nums[lp] + nums[rp] + n > 0:
                    rp -= 1
                else:
                    res.append([nums[lp], nums[rp], n])
                    lp += 1
                    rp -= 1
                    while lp < rp and nums[lp] == nums[lp - 1]:
                        lp += 1
                    
        return res
    

File: Data Structures & Algorithms/top-k-elements-in-list/submission-10.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> freq; // key (number) -> val (freq)
        for (int n : nums) {
            freq[n]++;
        }
        vector<pair<int, int>> res;
        for (const auto& [key, value] : freq){
            res.push_back({value, key});
        }
        sort(res.begin(), res.end());
        
        vector<int> final;
        for (int i = res.size() - 1; i >= (int)res.size() - k; i--){
            final.push_back(res[i].second);
        }
        return final;
    }
};
